2016/proposals/rai-vivek.md

```diff
@@ -0,0 +1,360 @@
+  > Using Flask web-development framework for Python. A robust, flexible
+  > choice that is easy to develop and maintain. Scalability is not an issue for
+  > our use case.
+
+* Server verifies the data, open database connection, updates the record.
+  > As suggested in the ideas sections, SQLite will be a low-barrier, easier to
+  > maintain database, appropriate for our needs.
+
+* Send a status response back to the client. The response can contain feedback,
+  suggestions or perhaps link to some survey.
+
+* Ability to export diagnostic data into JSON or CSV format.
+  > It’ll allow us to separate the user interface from backend creating a modular
+  > setup.
+
+#### Database
+
+Based on the current nature of collectable information, write-volume and
+functionality required SQLite seems to be a fit choice. In this regard, the
+database schema is an important component that determines how we are going to
+store the collected data in the database.
+
+The following illustrates a typical schema that might be used:
+
+```
+TABLE machine_info {
+  id PRIMARY KEY AUTO,
+  system VARCHAR,
+  hostname VARCHAR,
+  release VARCHAR,
+  version VARCHAR,
+  machine VARCHAR,
+  processor VARCHAR,
+  email VARCHAR,
+  time VARCHAR,
+  exit_status INT
+}
+
+
+TABLE software_version {
+  id PRIMARY KEY AUTO,
+  software VARCHAR,
+  installed BOOL,
+  version VARCHAR,
+  log VARCHAR
+}
+```
+
+Since we are obtaining the platform information using Python `platform` module
+which is same across all platforms, the same database schema can accommodate all
+systems.
+
+A typical example of the database entry then may look like:
+
+```
+(TABLE machine_info)
+
+id | system  | hostname    | release             | version                                               | machine  | processor | email       | time                      | exit_status
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+1  | ‘Linux’ | ‘ghostbook’ | ‘3.19.0-47-generic’ | ‘#53~14.04.1-Ubuntu SMP Mon Jan 18 16:09:14 UTC 2016’ | ‘x86_64’ | ‘x86_64’  | foo@bar.com | ‘Tue Mar 8 00:52:35 2016’ | 0
+
+(TABLE software_version)
+
+id | software | installed | version | log
+-----------------------------------------
+1  | python   | True      | 3.5.1   | ‘’
+1  | git      | True      | 1.9.1   | ‘’
+1  | hg       | False     | ‘’      | ‘’
+1  | pip      | True      | 7.9     | ‘A newer version of pip is available.’
+```
+
+Complex queries can be performed by `JOINING` the two tables and querying for
+corresponding field of interest.
+
+#### Concerns
+Both tables share the unique identity given to a particular report.  There is
+a possibility that same user may submit a report twice possibly because there
+were some errors or something else causing duplication.
```

Comment:
Okay, don't share the ID ever.

See: https://github.com/numfocus/gsoc/pull/94/files#r55733293
Just make it a 1-to-many relation.